AC-5421: First draft of the OrderedQueues heuristic

mc-nathan opened the pull request:
To test:
- `python3 allocate.py example.csv ordered_queues`

Run the results through analyze.py.  See the best results yet.

"missed Home Program Reads: 1" is due to the startup that has a program value of "Unknown Program".  It would be nice if the analyzer made this a bit clearer.

This PR does not complete this ticket since the universal queues and the reassignment queues are not yet implemented.

mc-nathan committed:
Initial version of Ordered Queues heuristic

mc-nathan committed:
Remove extra print statement

mc-nathan committed:
Merge branch 'development' into AC-5421

mc-nathan committed:
Removed unused local

mc-nathan committed:
Satisfy the CodeClimate

mc-nathan committed:
Readd heuristic specific assess

mc-nathan committed:
Merge branch 'AC-5519' into AC-5421

mc-nathan committed:
[AC-5421] Moved new classes into app_allocator

mc-nathan committed:
Remove blank line in classes/__init__.py

mc-nathan committed:
Merge branch 'development' into AC-5421

mc-nathan committed:
Fix imports for app_allocator

jonkiparsky commented on app_allocator/classes/specific_feature.py:
```diff
@@ -0,0 +1,53 @@
+from app_allocator.classes.option_spec import OptionSpec
+
+
+class SpecificFeature(Feature):
```
On their own, "Specific" and "Universal" don't really communicate the intent of these divisions well, and I can't think of a better self-documenting name. Docstrings for each of these classes (Feature, SpecificFeature, and UniversalFeature) would be useful to clarify this.

jonkiparsky commented on app_allocator/classes/specific_feature.py:
```diff
@@ -0,0 +1,53 @@
+# of the count in options1 divided by the count in options2.
+def sort_options(options1, options2):
+    options = set(options1.keys()).intersection(options2.keys())
+    weighted_options = [(options1[option]/float(options2[option]),
```
It is not necessary to coerce the divisor to a float in python3

mc-nathan replied:
Removed.

jonkiparsky commented on app_allocator/classes/specific_feature.py:
```diff
@@ -0,0 +1,53 @@
+    options = set(options1.keys()).intersection(options2.keys())
+    weighted_options = [(options1[option]/float(options2[option]),
+                         option) for option in options]
+    return [OptionSpec(option) for _, option in sorted(weighted_options)]
```
If you're sorting tuples, I would like to see a sort key, since it's not obvious to me how tuples sort in python or how you're expecting them to sort. Presumably you want them sorted by their zero element, so `sorted(opts, key = lambda tup: tup[0])` would explicitly specify that behavior. 

namedtuples might be worth considering for clarity, could probably be used to replace that comment.

mc-nathan replied:
Done.  namedtuples seem unnecessary to me.

jonkiparsky commented on app_allocator/classes/specific_feature.py:
```diff
@@ -0,0 +1,53 @@
+# options with counts.  E.g., {"Israel": 100, "Boston": 200}
+# The result is the set of shared options ordered by the ratio
+# of the count in options1 divided by the count in options2.
+def sort_options(options1, options2):
```
I don't love this name, since it is inherently ambiguous and incomplete. The ambiguity is "Do we mean that we're getting a set of options that are used for informing some kind of sort, or that we've got some options and we want them in order?" This can be resolved by inspection, but it breaks the flow when reading the code. 
The incompleteness is more significant, since this function is not just sorting options, it's also ensuring that the result contains only the options shared between the two sets. 
Something like `sorted_shared_options` might be a better name, or possibly thinking about the higher-level purpose and finding a name there. (ie, What is this in aid of?) Perhaps `weighted_options`?

mc-nathan replied:
Went for `_shared_options_by_scarcity`.

jonkiparsky commented on app_allocator/classes/specific_feature.py:
```diff
@@ -0,0 +1,53 @@
+# The result is the set of shared options ordered by the ratio
+# of the count in options1 divided by the count in options2.
+def sort_options(options1, options2):
+    options = set(options1.keys()).intersection(options2.keys())
```
considering the above, this set could be named `shared_options`

mc-nathan replied:
ok

jonkiparsky commented on app_allocator/classes/ordered_queues.py:
```diff
@@ -0,0 +1,123 @@
+        self.add_startups(startups)
+
+    def add_queues(self, slots):
+        # slots is an OrderDicts of possible options for each feature
```
s/b OrderedDict?

jonkiparsky commented on app_allocator/classes/queue.py:
```diff
@@ -0,0 +1,70 @@
+from app_allocator.classes.assignments import has_been_assigned
+
+
+class Queue(object):
```
Since Queue is both a standard generic data structure and a class in the standard library (https://docs.python.org/2/library/queue.html), and this is not a generic Queue (it's about judges and startups and assignments), the name should be more specific. Perhaps `AssigneeQueue` or `FeatureQueue`?

jonkiparsky commented on app_allocator/classes/queue.py:
```diff
@@ -0,0 +1,70 @@
+            return None
+        index = 0
+        count = len(self.items)
+        while index < count:
```
It would be more idiomatic to iterate over the objects in the list. 
```
for startup in self.items:
  if not self.has_been_assigned(judge, startup):
    ...
```

mc-nathan replied:
Done.

jonkiparsky commented on allocate.py:
```diff
@@ -24,5 +24,6 @@
 allocator.read_entities()
 allocator.setup()
 allocator.allocate()
+allocator.assess()
```
It's not clear to me why this is being put back, since we have a class to do analysis now.

mc-nathan replied:
So the heuristic objects have an opportunity to report useful information (like which queues still have stuff in them).

jonkiparsky commented on app_allocator/classes/option_spec.py:
```diff
@@ -0,0 +1,4 @@
+class OptionSpec(object):
```
Not a blocker, but this sort of minimal class seems like the sort of thing that the `namedtuple` was created for.

mc-nathan replied:
This might change soon, so I'm not going to change it, but point taken.  May change later.

jonkiparsky replied:
👍

jonkiparsky reviewed (changes requested):
This is my initial feedback, I have not yet absorbed all of this but this seems like a good chunk of stuff to consider. 
There may be more comments on this later.

mc-nathan committed:
[AC-5421] Add field specific queues
- Not sure this is the right approach, but checking it in case it is.

mc-nathan committed:
[AC-5421] Revised OrderedQueue heuristic
- Runs, but has some issues.  Commiting to integrate other review feedback.

mc-nathan committed:
Name changes for greater clarity

mc-nathan committed:
Universal replacement of startup with application

mc-nathan committed:
Merge branch 'development' into AC-5421

mc-nathan committed:
Merge branch 'development' into AC-5421

mc-nathan committed:
Replace startup with application after merge

mc-nathan committed:
Typo

mc-nathan committed:
process_judge_events should get a list of events

mc-nathan committed:
Rename SpecificFeature/UniversalFeature to MatchingFeature/JudgeFeature

mc-nathan committed:
code-check fixes

mc-nathan committed:
Coverage from OrderedQueues

mc-nathan committed:
Fix code-check

mc-nathan committed:
test_assess_* refactor

mc-nathan committed:
CodeClimate inspired refactors

mc-nathan committed:
Change Queue to NeedsQueue

mc-nathan committed:
Working OrderedQueues
- Fix accounting and queuing logic bugs
- Improve queue weighting system to include assignments and reads
- Handle reads through ReadCounter object
- Target 4 reads per app
- Start to generalize Features to improve read handling
- Analyzer provides example of maximum read case
- Tighten NeedsQueue, FieldNeed and OptionState printed representations
- Make judge assigned events more detailed
- Provide better interface for NeedsQueue

mc-nathan committed:
Reads as a Features
Use deepcopy method rather than rolling our own
Move assignment impact to Need objects
Read read_counter module

mc-nathan committed:
Get rid of queue with no needs

jonkiparsky commented on app_allocator/classes/read_need.py:
```diff
@@ -0,0 +1,25 @@
+            return self.count == other.count
+        return NotImplemented  # pragma: nocover
+
+    def process_action(self, action, _):
```
probably better to use `*args` if you need to consume arguments

mc-nathan committed:
Switch to priority order

mc-nathan committed:
Adjust CHANCE_OF_PASS to be more accurate based on data in example.csv

mc-nathan committed:
Add zscores

mc-nathan committed:
Support for Z-scores

mc-nathan committed:
More z-score changes

mc-nathan committed:
utils module for z-score calculations

mc-nathan committed:
Hide .DS_Store

mc-nathan committed:
code-check

mc-nathan committed:
Get rid of _ as an argument even when not used

mc-nathan committed:
Add deeper scenarios for OrderedQUeues testing

mc-nathan committed:
Make z-score aware of assignments

mc-nathan committed:
code-check and removed dead file

mc-nathan committed:
Refactor _find_best_queue

jonkiparsky commented on app_allocator/classes/assignments.py:
```diff
@@ -1,9 +1,59 @@
+    count = application.zscore_count
+    if count == 0:
+        return True
+    expected = application.expected_zscore(judge.zscore())
```
can this line be pushed down into `zscore_check`? This would allow us to just pass `application` and `judge` as parameters and let `zscore_check` handle the mechanics of how it does what it does.

mc-nathan replied:
As I see it the point of this function is to process the judge and application information so that `zscore_check` is just looking at numeric values.

jonkiparsky commented on app_allocator/classes/assignments.py:
```diff
@@ -1,9 +1,59 @@
+def can_be_assigned(judge, application):
+    if application in judge.all_applications:
+        return False
+    count = application.zscore_count
```
This should probably be pushed down into `zscore_check` as well since it's really just a special case of `COUNT_ZSCORE_LOOKUP`

mc-nathan replied:
From logical standpoint, if there are no values to base the application zscore on, does it make sense to compute the expected_zscore?  The logic for this code as stated by John H. clearly states if we don't have any judging data yet, then we take any judge regardless of their zscore.

Also conceptually, expected_zscore and estimated_zscore are the heaviest part of this so if we can avoid computing them that seems like a win.  In the current case they are trivial, but they could become more complicated, even involved database queries.

jonkiparsky replied:
So here's what I'm thinking - modified it a little to avoid the extra computations
This has not been tested, so modify as needed to account for any blunders. 

```
def can_be_assigned(judge, application):
    if application in judge.all_applications:
        return False
    return zscore_check(application, judge)

def zscore_check(application, judge):
    count = application.zscore_count
    if count == 0:
        return True
    expected = application.expected_zscore(judge.zscore())
    current = application.estimated_zscore()
    count_fn = COUNT_ZSCORE_LOOKUP.get(count, additional_reads)
    return count_fn(current, expected)
```

The main thing for me is that as written, `can_be_assigned` feels cluttered up logic that feels like it's just supporting `zscore_check`.

mc-nathan replied:
Seems like a matter of personal taste.  To me the above feels like zscore_check isn't really checking a Z-score and all can_be_assigned is doing is a single guard check.

jonkiparsky commented on app_allocator/classes/field_need.py:
```diff
@@ -0,0 +1,41 @@
+        self.option_states = option_states
+
+    def __str__(self):
+        return "F(%s, [%s])" % (self.field,
```
What is F?

mc-nathan replied:
Used to be "Field", but I was trying to make the output more concise so I could see a queue in a single line.

jonkiparsky replied:
okay, not a big deal (we will improve it later if it is ever actually used)

jonkiparsky commented on app_allocator/classes/needs_queue.py:
```diff
@@ -0,0 +1,78 @@
+    def work_left(self):
+        return len(self.items)
+
+    def judge_value(self, judge):
```
This is a little bit peculiar. It seems wrong to me that the value of this queue depends on the state of the first application in that queue, when there's no particular effort to ensure that the first app in the queue is the neediest. 
One possible error state that this could produce would be if the first app in the queue ends up with a low value for `value`, and there are needier applications behind it. Those applications would then be blocked from getting reads until that relatively needless app got selected. 
How do we guard against that condition?

Not a blocker, since this is not production code, but I wanted to make sure it was noted.

mc-nathan replied:
It's not based on the first app in the queue.  It's based on the application this judge would get if we take an application from this queue.  That seems like exactly the right thing to be evaluating.  If their are needier apps behind the one that was chosen, then that's an issue with `_next_application`.  Currently this heuristic does just look for the first application that meets the z-score criteria and that the judge has not judged.  If the application at the front of the queue has no assignments in flight, then the first one is fine.  The tricky part is when the front element has been assigned.  It then becomes a question of how hard we want to look for a better match.  If we look exhaustively, then there's significantly less benefit to having the queues.  One option would be to figure out for each queue what the greatest possible value is (which would be for an unassigned application) and not search queues whose greatest possible value is less than the value of the current candidate.  There would still be cases where you would have to do an exhaustive search to find the "best".  I think solving this question is outside the scope of this PR.

jonkiparsky replied:
I agree that the reassignment condition is where the potential problem arises, not that concerned about whether the "real" problem is here or in `_next_application` - it arises most clearly here, this is where we'd see the bug if it arises, so this is where I mentioned it. 

I agree that the solution is not required for this PR, hence not a blocker, but it seems like a solvable problem that is worth solving.

jonkiparsky commented on app_allocator/classes/ordered_queues.py:
```diff
@@ -0,0 +1,161 @@
+            else:
+                Event(action="complete",
+                      subject=queue)
+#         self.assess_zscore()
```
commented code - can be removed?

mc-nathan replied:
This is Z-score debugging code that I'm not quite ready to let go.

mc-nathan replied:
Switched it to a boolean.

jonkiparsky commented on app_allocator/classes/read_need.py:
```diff
@@ -0,0 +1,25 @@
+        self.count = count
+
+    def __str__(self):
+        return "R(%s)" % self.count
```
I don't really love this convention. Could we maybe use words?

mc-nathan replied:
Again this was really a debugging thing so I could comprehend the output.

jonkiparsky replied:
not a big deal - if it's more convenient this way that's fine

jonkiparsky reviewed (changes requested):
There are a few minor points that seem worth looking at.

mc-nathan committed:
Added control for controlling z-score reporting info
